src/consensus/pbft/pbft.h

```diff
@@ -73,11 +85,74 @@ namespace pbft
         reinterpret_cast<const uint8_t*>(msg->contents()),
         msg->size());
 
+      if (msg->tag() == Append_entries_tag)
+      {
+        auto node = nodes.find(to);
+
+        Index match_idx = 0;
+        if (node != nodes.end())
+        {
+          match_idx = node->second;
+        }
+
+        if (match_idx < latest_stable_ae_index)
+        {
+          send_append_entries(to, match_idx + 1);
+        }
+        return msg->size();
+      }
       n2n_channels->send_authenticated(
         ccf::NodeMsgType::consensus_msg, to, serialized_msg);
       return msg->size();
     }
 
+    void send_append_entries(NodeId to, Index start_idx)
+    {
+      Index end_idx = (latest_stable_ae_index == 0) ?
+        0 :
+        std::min(start_idx + entries_batch_size, latest_stable_ae_index);
+
+      for (Index i = end_idx; i < latest_stable_ae_index;
+           i += entries_batch_size)
+      {
+        send_append_entries_range(to, start_idx, i);
+        start_idx = std::min(i + 1, latest_stable_ae_index);
+      }
+
+      if (latest_stable_ae_index == 0 || end_idx <= latest_stable_ae_index)
+      {
+        send_append_entries_range(to, start_idx, latest_stable_ae_index);
+      }
+    }
+
+    void send_append_entries_range(NodeId to, Index start_idx, Index end_idx)
+    {
+      const auto prev_idx = start_idx - 1;
+
+      LOG_INFO_FMT(
+        "Send append entried from {} to {}: {} to {}",
+        id,
+        to,
+        start_idx,
+        end_idx);
+
+      AppendEntries ae = {pbft_append_entries, id, end_idx, prev_idx};
+
+      auto node = nodes.find(to);
+      if (node != nodes.end())
```

Comment:
As discussed, the catching up node's "match index" should only be updated only once this node has received the AE response. Otherwise, assuming unreliable transport, the catching up node may not receive AE and will forever be stuck behind.

Reply:
Right now we don't have append entries response implemented since it means either sending out a new message along with every status message, or intercepting any outcoming status messages from `Send()`, setting the AE index, and then sending it. On the incoming side it means checking if an incoming message is a status one, deserialising it, getting the append entries "response" and then forwarding it into the Replica to be processed. 

This will work fine for now without AE responses, since _all_ replicas will send append entries to late joiners and it is highly unlikely that a late joiner will be stuck behind for ever. I think this can be implemented later as part of #753